app/src/main/java/com/danosoftware/galaxyforce/services/googleplay/GooglePlayServices.java

```diff
@@ -0,0 +1,548 @@
+            source.setResult(result.getData());
+            return source.getTask();
+        }
+
+        // There was a conflict - we need to resolve it.
+        SnapshotsClient.SnapshotConflict conflict = result.getConflict();
+
+        Snapshot snapshot = conflict.getSnapshot();
+        Snapshot conflictSnapshot = conflict.getConflictingSnapshot();
+        GooglePlaySavedGame savedGame = extractSavedGame(snapshot);
+        GooglePlaySavedGame conflictSavedGame = extractSavedGame(conflictSnapshot);
+
+        // Resolve between conflicts by selecting the snapshots with the most progress.
+        Snapshot resolvedSnapshot = snapshot;
+        if (savedGame != null
+                && conflictSavedGame != null
+                && conflictSavedGame.getHighestWaveReached() > savedGame.getHighestWaveReached()) {
+            resolvedSnapshot = conflictSnapshot;
+        } else if (savedGame == null
+                && conflictSavedGame != null) {
+            resolvedSnapshot = conflictSnapshot;
+        }
+
+        return Games.getSnapshotsClient(mActivity, GoogleSignIn.getLastSignedInAccount(mActivity))
+                .resolveConflict(conflict.getConflictId(), resolvedSnapshot)
+                .continueWithTask(
+                        new Continuation<
+                                SnapshotsClient.DataOrConflict<Snapshot>,
+                                Task<Snapshot>>() {
+                            @Override
+                            public Task<Snapshot> then(
+                                    @NonNull Task<SnapshotsClient.DataOrConflict<Snapshot>> task)
+                                    throws Exception {
+                                // Resolving the conflict may cause another conflict,
+                                // so recurse and try another resolution.
+                                if (retryCount < MAX_SNAPSHOT_RESOLVE_RETRIES) {
+                                    return processSnapshotAndResolveConflictsTask(task.getResult(), retryCount + 1);
+                                } else {
+                                    throw new GalaxyForceException("Could not resolve snapshot conflicts");
+                                }
+                            }
+                        });
+    }
+
+
+    /**
+     * Returns a task to write the save game snapshot with a suitable description.
+     */
+    private Task<SnapshotMetadata> writeSnapshotTask(
+            Snapshot snapshot,
+            GooglePlaySavedGame savedGame) throws JsonProcessingException {
+
+        // Set the data payload for the snapshot
+        final ObjectMapper mapper = new ObjectMapper();
+        final byte[] gameData = mapper.writeValueAsBytes(savedGame);
+        snapshot.getSnapshotContents().writeBytes(gameData);
+
+        final int waveUnlocked = savedGame.getHighestWaveReached();
+        final boolean finishedAllWaves = waveUnlocked > GameConstants.MAX_WAVES;
```

Comment:
Is it possible that waveUnlocked will ever exceed `MAX_WAVES`. Logic may prevent this.